File: src/risk.ts
export interface PerpsMarketView {
  name: string;
  maxLeverage: number;
}

export interface PerpsPositionView {
  id: string;
  market: string;
  side: "long" | "short";
  size: number;
  leverage: number;
  entryPrice: number;
  markPrice: number;
  margin: number;
  unrealizedPnl: number;
  liquidationPrice: number;
}

export interface PositionRisk {
  id: string;
  market: string;
  side: "long" | "short";
  size: number;
  leverage: number;
  markPrice: number;
  liquidationPrice: number | null;
  notionalUsd: number;
  marginUsd: number;
  unrealizedPnlUsd: number;
  pnlOnMarginPct: number | null;
  liquidationDistancePct: number | null;
  leverageUtilizationPct: number | null;
  warnings: string[];
}

export interface PerpsRiskSnapshot {
  address: string;
  computedAt: string;
  warnWithinPct: number;
  positionCount: number;
  totals: {
    notionalUsd: number;
    marginUsd: number;
    unrealizedPnlUsd: number;
  };
  nearestLiquidationDistancePct: number | null;
  warningCount: number;
  positions: PositionRisk[];
}

function round(value: number, digits = 6): number {
  const scale = 10 ** digits;
  return Math.round((value + Number.EPSILON) * scale) / scale;
}

export function validateWarningThreshold(value: number): number {
  if (!Number.isFinite(value) || value <= 0) {
    throw new Error("--warn-within must be a positive percentage");
  }
  return value;
}

export function buildPositionRisk(
  position: PerpsPositionView,
  market: PerpsMarketView | undefined,
  warnWithinPct: number,
): PositionRisk {
  validateWarningThreshold(warnWithinPct);

  const warnings: string[] = [];
  const markPrice = Number.isFinite(position.markPrice) ? position.markPrice : 0;
  const liquidationPrice =
    Number.isFinite(position.liquidationPrice) && position.liquidationPrice > 0
      ? position.liquidationPrice
      : null;

  const notionalUsd = Math.abs(position.size) * Math.max(markPrice, 0);
  const pnlOnMarginPct =
    Number.isFinite(position.margin) && position.margin > 0
      ? (position.unrealizedPnl / position.margin) * 100
      : null;

  let liquidationDistancePct: number | null = null;
  if (liquidationPrice === null) {
    warnings.push("liquidation price unavailable");
  } else if (markPrice <= 0) {
    warnings.push("mark price unavailable; liquidation buffer cannot be computed");
  } else {
    liquidationDistancePct =
      position.side === "long"
        ? ((markPrice - liquidationPrice) / markPrice) * 100
        : ((liquidationPrice - markPrice) / markPrice) * 100;

    if (liquidationDistancePct <= 0) {
      warnings.push("liquidation price is at or beyond the reported mark");
    } else if (liquidationDistancePct <= warnWithinPct) {
      warnings.push(
        `liquidation buffer ${round(liquidationDistancePct, 2)}% is within ${round(warnWithinPct, 2)}% threshold`,
      );
    }
  }

  let leverageUtilizationPct: number | null = null;
  if (!market || !Number.isFinite(market.maxLeverage) || market.maxLeverage <= 0) {
    warnings.push("market leverage metadata unavailable");
  } else {
    leverageUtilizationPct = (position.leverage / market.maxLeverage) * 100;
    if (leverageUtilizationPct > 100) {
      warnings.push(
        `reported leverage ${position.leverage}x exceeds returned market maximum ${market.maxLeverage}x`,
      );
    }
  }

  return {
    id: position.id,
    market: position.market,
    side: position.side,
    size: position.size,
    leverage: position.leverage,
    markPrice: round(markPrice),
    liquidationPrice: liquidationPrice === null ? null : round(liquidationPrice),
    notionalUsd: round(notionalUsd),
    marginUsd: round(position.margin),
    unrealizedPnlUsd: round(position.unrealizedPnl),
    pnlOnMarginPct: pnlOnMarginPct === null ? null : round(pnlOnMarginPct),
    liquidationDistancePct:
      liquidationDistancePct === null ? null : round(liquidationDistancePct),
    leverageUtilizationPct:
      leverageUtilizationPct === null ? null : round(leverageUtilizationPct),
    warnings,
  };
}

export function buildRiskSnapshot(options: {
  address: string;
  positions: PerpsPositionView[];
  markets: PerpsMarketView[];
  warnWithinPct: number;
  computedAt?: string;
}): PerpsRiskSnapshot {
  const warnWithinPct = validateWarningThreshold(options.warnWithinPct);
  const marketByName = new Map(options.markets.map((market) => [market.name, market]));
  const positions = options.positions.map((position) =>
    buildPositionRisk(position, marketByName.get(position.market), warnWithinPct),
  );
  const distances = positions
    .map((position) => position.liquidationDistancePct)
    .filter((value): value is number => value !== null);

  return {
    address: options.address,
    computedAt: options.computedAt ?? new Date().toISOString(),
    warnWithinPct: round(warnWithinPct),
    positionCount: positions.length,
    totals: {
      notionalUsd: round(positions.reduce((sum, position) => sum + position.notionalUsd, 0)),
      marginUsd: round(positions.reduce((sum, position) => sum + position.marginUsd, 0)),
      unrealizedPnlUsd: round(
        positions.reduce((sum, position) => sum + position.unrealizedPnlUsd, 0),
      ),
    },
    nearestLiquidationDistancePct: distances.length ? round(Math.min(...distances)) : null,
    warningCount: positions.reduce((sum, position) => sum + position.warnings.length, 0),
    positions,
  };
}

File: risk.py
"""Pure, read-only risk metrics for the Suwappu perps example."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
import math
from typing import Mapping, Sequence


@dataclass(frozen=True)
class PositionInput:
    id: str
    market: str
    side: str
    size: float
    leverage: float
    entry_price: float
    mark_price: float
    margin: float
    unrealized_pnl: float
    liquidation_price: float
    funding_rate: float | None = None


def _rounded(value: float, digits: int = 6) -> float:
    return round(value, digits)


def validate_warning_threshold(value: float) -> float:
    if not math.isfinite(value) or value <= 0:
        raise ValueError("--warn-within must be a positive percentage")
    return value


def build_position_risk(
    position: PositionInput,
    market_max_leverage: float | None,
    warn_within_pct: float,
) -> dict[str, object]:
    validate_warning_threshold(warn_within_pct)
    warnings: list[str] = []

    mark_price = position.mark_price if math.isfinite(position.mark_price) else 0.0
    liquidation_price = (
        position.liquidation_price
        if math.isfinite(position.liquidation_price) and position.liquidation_price > 0
        else None
    )
    notional_usd = abs(position.size) * max(mark_price, 0.0)
    pnl_on_margin_pct = (
        position.unrealized_pnl / position.margin * 100
        if math.isfinite(position.margin) and position.margin > 0
        else None
    )

    liquidation_distance_pct: float | None = None
    if liquidation_price is None:
        warnings.append("liquidation price unavailable")
    elif mark_price <= 0:
        warnings.append("mark price unavailable; liquidation buffer cannot be computed")
    else:
        if position.side == "long":
            liquidation_distance_pct = (mark_price - liquidation_price) / mark_price * 100
        else:
            liquidation_distance_pct = (liquidation_price - mark_price) / mark_price * 100

        if liquidation_distance_pct <= 0:
            warnings.append("liquidation price is at or beyond the reported mark")
        elif liquidation_distance_pct <= warn_within_pct:
            warnings.append(
                f"liquidation buffer {_rounded(liquidation_distance_pct, 2):g}% "
                f"is within {_rounded(warn_within_pct, 2):g}% threshold"
            )

    leverage_utilization_pct: float | None = None
    if (
        market_max_leverage is None
        or not math.isfinite(market_max_leverage)
        or market_max_leverage <= 0
    ):
        warnings.append("market leverage metadata unavailable")
    else:
        leverage_utilization_pct = position.leverage / market_max_leverage * 100
        if leverage_utilization_pct > 100:
            warnings.append(
                f"reported leverage {position.leverage:g}x exceeds returned market maximum "
                f"{market_max_leverage:g}x"
            )

    return {
        "id": position.id,
        "market": position.market,
        "side": position.side,
        "size": position.size,
        "leverage": position.leverage,
        "markPrice": _rounded(mark_price),
        "liquidationPrice": (
            None if liquidation_price is None else _rounded(liquidation_price)
        ),
        "notionalUsd": _rounded(notional_usd),
        "marginUsd": _rounded(position.margin),
        "unrealizedPnlUsd": _rounded(position.unrealized_pnl),
        "fundingRate": (
            None
            if position.funding_rate is None or not math.isfinite(position.funding_rate)
            else _rounded(position.funding_rate)
        ),
        "pnlOnMarginPct": (
            None if pnl_on_margin_pct is None else _rounded(pnl_on_margin_pct)
        ),
        "liquidationDistancePct": (
            None
            if liquidation_distance_pct is None
            else _rounded(liquidation_distance_pct)
        ),
        "leverageUtilizationPct": (
            None
            if leverage_utilization_pct is None
            else _rounded(leverage_utilization_pct)
        ),
        "warnings": warnings,
    }


def build_risk_snapshot(
    address: str,
    positions: Sequence[PositionInput],
    market_max_by_name: Mapping[str, float],
    warn_within_pct: float,
    computed_at: str | None = None,
) -> dict[str, object]:
    warn_within_pct = validate_warning_threshold(warn_within_pct)
    position_risks = [
        build_position_risk(
            position,
            market_max_by_name.get(position.market),
            warn_within_pct,
        )
        for position in positions
    ]
    distances = [
        value
        for item in position_risks
        if (value := item["liquidationDistancePct"]) is not None
    ]

    return {
        "address": address,
        "computedAt": computed_at
        or datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace(
            "+00:00", "Z"
        ),
        "warnWithinPct": _rounded(warn_within_pct),
        "positionCount": len(position_risks),
        "totals": {
            "notionalUsd": _rounded(
                sum(float(item["notionalUsd"]) for item in position_risks)
            ),
            "marginUsd": _rounded(
                sum(float(item["marginUsd"]) for item in position_risks)
            ),
            "unrealizedPnlUsd": _rounded(
                sum(float(item["unrealizedPnlUsd"]) for item in position_risks)
            ),
        },
        "nearestLiquidationDistancePct": (
            _rounded(min(float(value) for value in distances)) if distances else None
        ),
        "warningCount": sum(len(item["warnings"]) for item in position_risks),
        "positions": position_risks,
    }
